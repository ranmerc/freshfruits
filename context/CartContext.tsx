import { useLocalStorage } from "usehooks-ts";
import {
  createContext,
  useContext,
  ReactNode,
  useReducer,
  Dispatch,
  useCallback,
} from "react";
import CartItem from "@/types/CartItem";

type ActionType =
  | {
      type: "ADD_TO_CART";
      payload: CartItem;
    }
  | {
      type: "REMOVE_FROM_CART";
      payload: {
        fruitId: number;
        selectedPackId: number;
      };
    }
  | {
      type: "CLEAR_CART";
    };

interface CartContextType {
  cartItems: CartItem[];
  cartDispatch: Dispatch<ActionType>;
}

const CartContext = createContext<CartContextType | undefined>(undefined);

export function useCartContext() {
  const context = useContext(CartContext);

  if (!context) {
    throw new Error("useCartContext must be used within a CartProvider");
  }

  return context;
}

function cartReducer(state: CartItem[], action: ActionType) {
  switch (action.type) {
    case "ADD_TO_CART": {
      const { fruitId, selectedPackId, quantity } = action.payload;

      const existingItemIndex = state.findIndex(
        (item) =>
          item.fruitId === fruitId && item.selectedPackId === selectedPackId
      );

      if (existingItemIndex !== -1) {
        const newState = [...state];
        newState[existingItemIndex].quantity += quantity;
        return newState;
      }

      return [...state, action.payload];
    }

    case "REMOVE_FROM_CART": {
      const { fruitId, selectedPackId } = action.payload;

      const existingItemIndex = state.findIndex(
        (item) =>
          item.fruitId === fruitId && item.selectedPackId === selectedPackId
      );

      if (existingItemIndex === -1) {
        return state;
      }

      const newState = [...state];
      newState[existingItemIndex].quantity -= 1;

      if (newState[existingItemIndex].quantity === 0) {
        newState.splice(existingItemIndex, 1);
      }

      return newState;
    }

    case "CLEAR_CART":
      return [];
    default:
      return state;
  }
}

export function CartProvider({ children }: { children: ReactNode }) {
  const [savedCartItems, setSavedCartItems] = useLocalStorage<CartItem[]>(
    "cart-items",
    []
  );

  const reducerLocalStorage = useCallback(
    (state: CartItem[], action: ActionType) => {
      const newState = cartReducer(state, action);

      setSavedCartItems(newState);

      return newState;
    },
    [setSavedCartItems]
  );

  const [state, dispatch] = useReducer(reducerLocalStorage, savedCartItems);

  return (
    <CartContext.Provider value={{ cartItems: state, cartDispatch: dispatch }}>
      {children}
    </CartContext.Provider>
  );
}
