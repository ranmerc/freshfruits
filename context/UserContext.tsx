import { createContext, useContext, ReactNode } from "react";
import { useLocalStorage } from "usehooks-ts";

export interface UserContextType {
  username: string;
  setUserName: (user: string) => void;
}

export const UserContext = createContext<UserContextType | undefined>(
  undefined
);

export function useUserContext() {
  const context = useContext(UserContext);

  if (!context) {
    throw new Error("useUserContext must be used within a UserProvider");
  }

  return context;
}

export function UserProvider({ children }: { children: ReactNode }) {
  const [username, setUserName] = useLocalStorage<string>("username", "");

  return (
    <UserContext.Provider value={{ username, setUserName }}>
      {children}
    </UserContext.Provider>
  );
}
