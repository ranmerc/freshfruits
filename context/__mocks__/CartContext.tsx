export const CartContext = jest.fn();

export const useCartContext = jest.fn();

export const CartProvider = jest.fn();
