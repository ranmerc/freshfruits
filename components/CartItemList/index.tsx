import { useCartContext } from "@/context/CartContext";
import CartItem from "@/components/CartItem";
import { Stack, Typography } from "@mui/material";

export default function CartItemList() {
  const { cartItems } = useCartContext();

  if (cartItems.length === 0) {
    return (
      <Typography variant={"body1"} mt={2}>
        Your cart is empty!
      </Typography>
    );
  }

  return (
    <>
      <Stack component={"ul"} sx={{ listStyle: "none", padding: 0, rowGap: 3 }}>
        {cartItems.map((item) => (
          <li key={item.fruitId}>
            <CartItem item={item} />
          </li>
        ))}
      </Stack>
    </>
  );
}
