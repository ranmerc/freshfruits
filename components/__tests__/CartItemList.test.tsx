import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import CartItemList from "../CartItemList";
import { useCartContext } from "@/context/CartContext";
import renderer from "react-test-renderer";

jest.mock("@/context/CartContext");
jest.mock("@/components/CartItem", () => {
  return function CartItem() {
    return <div>Cart Item</div>;
  };
});

const mockedUseCartContext = useCartContext as jest.Mock;

describe("CartItemList component", () => {
  it("Renders correctly when list is empty", () => {
    mockedUseCartContext.mockReturnValue({ cartItems: [] });
    render(<CartItemList />);

    expect(screen.getByRole("alert")).toHaveTextContent("Your cart is empty!");

    const tree = renderer.create(<CartItemList />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("Renders list", () => {
    const items = [
      {
        fruitId: 1,
        selectedPackId: 1,
      },
      {
        fruitId: 2,
        selectedPackId: 1,
      },
    ];

    mockedUseCartContext.mockReturnValue({ cartItems: items });
    render(<CartItemList />);

    expect(screen.getByRole("list")).toBeInTheDocument();
    expect(screen.getAllByRole("listitem")).toHaveLength(2);

    const tree = renderer.create(<CartItemList />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
