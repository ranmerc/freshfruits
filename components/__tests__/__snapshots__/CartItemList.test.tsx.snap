// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItemList component Renders correctly when list is empty 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardInfo MuiAlert-standard css-3kfj0h-MuiPaper-root-MuiAlert-root"
  role="alert"
>
  <div
    className="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
      data-testid="InfoOutlinedIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20, 12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10, 10 0 0,0 12,2M11,17H13V11H11V17Z"
      />
    </svg>
  </div>
  <div
    className="MuiAlert-message css-1pxa9xg-MuiAlert-message"
  >
    Your cart is empty!
  </div>
</div>
`;

exports[`CartItemList component Renders list 1`] = `
<ul
  className="MuiStack-root css-1k2rdr-MuiStack-root"
>
  <li>
    <div>
      Cart Item
    </div>
  </li>
  <li>
    <div>
      Cart Item
    </div>
  </li>
</ul>
`;
